src/collapse/common.rs

```diff
@@ -0,0 +1,780 @@
+use crossbeam::channel;
+use fnv::FnvHashMap;
+use lazy_static::lazy_static;
+
+const CAPACITY_HASHMAP: usize = 512;
+pub(crate) const CAPACITY_READER: usize = 128 * 1024;
+pub(crate) const DEFAULT_NSTACKS_PER_JOB: usize = 100;
+const NBYTES_PER_STACK_GUESS: usize = 1024;
+const RUST_HASH_LENGTH: usize = 17;
+
+lazy_static! {
+    #[doc(hidden)]
+    pub static ref DEFAULT_NTHREADS: usize = num_cpus::get();
+}
+
+/// Private trait for internal library authors.
+///
+/// If you implement this trait, your type will implement the public-facing
+/// `Collapse` trait as well. Implementing this trait gives you parallelism
+/// for free as long as you adhere to the requirements described in the
+/// comments below.
+pub trait CollapsePrivate: Clone + Send + Sized + Sized {
+    // *********************************************************** //
+    // ********************* REQUIRED METHODS ******************** //
+    // *********************************************************** //
+
+    /// Some formats, such as `dtrace`, contain a header or other non-stack
+    /// information at the beginning of their input files. If header information
+    /// is present, this method **must** consume it (i.e. advance the provided
+    /// reader past it).
+    ///
+    /// This method also provides an opportunity to do processing of actual
+    /// stack data on the main thread before worker threads are spun up. For
+    /// an example of why this might be necessary, see `perf`, which can require
+    /// reading the first stack in order to know how to process the rest.
+    ///
+    /// If the format you are working with does not contain header information
+    /// or does not need any special, up-front processing, just have this method
+    /// return `Ok(())` immediately.
+    fn pre_process<R>(&mut self, reader: &mut R, occurrences: &mut Occurrences) -> io::Result<()>
+    where
+        R: io::BufRead;
+
+    /// The primary method.
+    ///
+    /// It receives a reader whose header has already been consumed (see above), as
+    /// well as a mutable reference to an `Occurences` instance (just a hashamp that
+    /// works across multiple threads). Implementators should parse the stack data
+    /// contained in the reader and write output to the provided `Occurrences` map.
+    ///
+    /// So that this method may be called multiple times, all internal
+    /// state contained in `self` (e.g. any stack buffers or the like) **must** be
+    /// reset before this method returns. Also, this method may **not** use threads.
+    fn collapse_single_threaded<R>(
+        &mut self,
+        reader: R,
+        occurrences: &mut Occurrences,
+    ) -> io::Result<()>
+    where
+        R: io::BufRead;
+
+    /// This method receives a line from the input data (including the trailing newline byte)
+    /// and should return `true` if the line represents the end of a stack; `false` otherwise.
+    fn is_end_of_stack(&self, line: &[u8]) -> bool;
```

Comment:
This _might_ be tricky for some implementations solely based on the current line. I guess since they have access to `&self`, the implementation might make a context-based decision, though I think the signature should reflect that. Perhaps:
```rust
fn would_end_stack(&self, line: &[u8]) -> bool;
```
I also think the documentation here should specify how this is used. For example, will this ever be called before any stacks have been parsed?

Reply:
I agree. My thought is that if another implementation requires tweaks to the algorithms used in `CollapsePrivate`, we can address those once we know what other implementations might need. I don't think we'll be able to fully anticipate every possible scenario until we see them. This implementation effectively abstracts across `perf` and `dtrace`. If we add another collapser later that needs slight changes to `CollapsePrivate` to work, let's make those changes then.

Reply:
That said, I did change the name of this method to `would_end_stack` as well as added more context to the documentation.

Reply:
Also - As you can see with how `sample` is now, collapsers don't have to implement `CollapsePrivate` if it doesn't work for them. They can always just implement `Collapse` itself.